Abstract out manifests more

apple1417 opened the pull request:
Paving the way to adding editor.

Changed `Datum` class to `SteamManifest`, which also holds app/depot id and package. Because of this switched the internal data format of `ManifestData` to a `Dictionary<int, List<SteamManifest>>`
Moved the initialization of the internal data into a private function `InitData()`, allowing the class to become static.
Removed the `[]` operator in favor of a function `GetManifestsForVersion()`, now that there's no longer a nested dict - package isn't an index in any sense.

apple1417 committed:
Abstract out manifests more

jbzdarkid commented on DepotManager.cs:
```diff
@@ -90,19 +87,20 @@ public class DepotManager {
         if (Settings.Default.ownsGehenna && !IsFullyDownloaded(version, Package.Gehenna)) {
-          neededDepots.Add(ManifestData.GEHENNA);
+          neededManifests.AddRange(ManifestData.GetManifestsForVersion(version, Package.Gehenna));
+          totalDownloadSize += ManifestData.GetDownloadSize(version, Package.Gehenna);
```
Having several calls to GetDownloadSize doesn't seem cleaner. Why not preserve the old loop?

apple1417 replied:
The old way ends up duplicating code. This was one of the spots I was thinking a flags enum would help a decent bit, though I ended up not going with it cause it makes getting a folder per package awkward, it's a bit of a remnant from when I was trying that too.

In any case I'll revert it.

jbzdarkid commented on ManifestData.cs:
```diff
@@ -23,251 +31,332 @@ public class ManifestData {
-        return this[version, PROTOTYPE].size;
-      } else {
-        return 0;
+    public static long GetDownloadSize(int version, Package package = Package.All) {
```
I'd like to avoid default parameters, unless there's a common use case. In this case, since it's not used at all, I'd like to avoid it (since it can lead to bugs / unexpected behavior).

jbzdarkid commented on SteamCommand.cs:
```diff
@@ -15,10 +15,12 @@ public static class SteamCommand {
-      if (manifest == 0) return; // 0 indicates "No such manifest", so we shouldn't attempt to download it.
-      sim.Keyboard.TextEntry($"download_depot {GAME_ID} {depot} {manifest}");
+    public static void DownloadManifest(SteamManifest manifest) {
+      if (manifest is null) throw new System.ArgumentNullException(nameof(manifest));
```
Don't throw exceptions, there's no exception handling. This will crash the app.
I suggest following the old idea, and simply electing to not download if there is no manifest.

Also, I'd prefer not using the `is` keyword, it's a bit too python-esque.

apple1417 replied:
Ah windows forms vs wpf must be different there, there's a window for uncaught exceptions in forms, so I figured it'd be fine to throw an exception that never happens anyway. Should I have it return if the data on the manifest is 0 too, like before?

jbzdarkid replied:
Even if there is a handler, I don't like showing a dialogue box that means nothing to the user.
I assumed you were replacing `ManifestData(0, 0, 0)` with `null`, so the `== 0` check is gone? If not, you should handle that case, too.

apple1417 replied:
In the cases where that used to be used I just don't include the manifests in the list, this was just a warning intelisense gave me once I got it working again.

jbzdarkid commented on ManifestData.cs:
```diff
@@ -23,251 +31,332 @@ public class ManifestData {
+    }
+
+    public static IEnumerable<SteamManifest> GetManifestsForVersion(int version, Package package = Package.All) {
+      return data[version].Where(m => package == Package.All || package == m.package);
```
This is a low-performance implementation. If you know you need a second key, use a second key.
How about `Dictionary<int, Dictionary<Package, List<SteamManifest>> data`?
That way you can access via `data[version][package]`.

jbzdarkid commented on ManifestData.cs:
```diff
@@ -23,251 +31,332 @@ public class ManifestData {
+    }
+
+    private static readonly Dictionary<int, List<SteamManifest>> data = InitData();
+    private static Dictionary<int, List<SteamManifest>> InitData() {
```
If you're going to make a singleton, make a singleton. Don't have another structure represent the singleton instance.
```C#
private static ManifestData instance = CreateInstance();
private static ManifestData CreateInstance() {
  if (instance != null) return;
  ...
}

public static long GetDownloadSize(...) {
  if (instance == null) return;
  lock(instance) {
    ...
  }
}
```

jbzdarkid replied:
I guess if it's a `readonly` instance, then you don't need to lock.

apple1417 replied:
Why would we want/need a singleton over a static class here? Not sure I fully understand the justification.

jbzdarkid replied:
I was thinking it would make things cleaner, since you have a "singleton" data class member, but on further thought, it's fine as-is.

jbzdarkid commented on Package.cs:
```diff
@@ -3,5 +3,6 @@ public enum Package {
     Main,
     Gehenna,
     Prototype,
+    All
```
I don't think this is necessary at the moment. Let's leave it out until you need it.

jbzdarkid commented on DepotManager.cs:
```diff
@@ -135,13 +133,8 @@ public class DepotManager {
-
-          CopyAndOverwrite($"{depotLocation}/depot_{depot}", GetFolder(version, package), delegate (long fileSize) {
+        foreach (var manifest in neededManifests) {
+          CopyAndOverwrite($"{depotLocation}/depot_{manifest.depotId}", GetFolder(version, manifest.package), delegate (long fileSize) {
```
`depotLocation` is currently hard-coded to the talos appId. Can you update it to use the value from the manifest now?

jbzdarkid commented on ManifestData.cs:
```diff
@@ -23,251 +31,332 @@ public class ManifestData {
+        new SteamManifest(257510, 322022, 4383443687812477256,  3,  206909511, Package.Prototype)
+      };
+
+      data[220996] = new List<SteamManifest> {
```
This new ordering is less helpful, since adding new packages is now an interleaved process, rather than a cumulative one. It also separates similarities, which in my experience makes copy/paste errors harder to spot. Is there any particular reason you felt the need to change this?

apple1417 replied:
With the old ordering adding a new version was an interleaved process instead. There isn't really a way to avoid it for both, one of the two has to be interleaved like that. Adding new versions is almost certainly going to be more common than packages, so I switched it.

Somewhat relatedly: I feel like it might be better to load all this data from a file, it becomes easier to update and it should be possible to check a url to add any new versions without updating the whole application. That goes beyond the scope of this PR though.

jbzdarkid replied:
There's only 5-10 packages, so adding a new version is only in 5-10 places.
However, you're planning on adding in editor packages, and there are >30 versions.

I think the long-term solution is storing it in a table so that we can easily re-sort the data. I think WPF has some sort of "data" format?

I don't know why you think a URL is necessary, given how infrequently this game is being updated. However, if you want to move this data table into another file, that is fine.

apple1417 replied:
Adding editor is a one time cost, as far as we know they'll keep updating the game indefinitely and we'll have to keep adding new versions, which is certainly much easier when everything is in one place.

Any sort of re-sortable data format also won't fix the actual "issue" here, the code layout. Looking at parsing it from a file though, I do think the same version based format would be better, it mirrors the same sort of structure as the internal dictionary - `data[<version>][<package>][<index>].attribute`.

For now I'll just commit the other fixes and leave this alone, but if it's really such an issue I'll change it in the next one, not like it's particularly hard to rearrange this stuff.

apple1417 committed:
Code review changes

Major change is that the ManifestData internal table is now a `Dictionary<int, Dictionary<Package, List<SteamManifest>>`

jbzdarkid commented:
@apple1417 I pulled in some of your changes, let me know if there's anything else in this PR that you think is needed.
